### ISSUE TITLE
Logical compound assignment operators &&= and ||=

### ISSUE THREAD

**Issue by eernstg** [dart-lang/language#23]
Cf. issues https://github.com/dart-lang/sdk/issues/26996, meta, and https://github.com/dart-lang/sdk/issues/26997, analyzer, (there are others, but they should probably be replaced by a single front-end issue), the compound assignment operators `&&=` and `||=` have been in the pipeline for a full specification and an implementation for a long time.

We ought to find a placement for this task in terms of milestones and priorities, or drop it.

**Comment by thosakwe** [comment]
The following prints out `true`:

```dart
void main() {
  var a = false;
  a |= true;
  print(a);
}
```
So maybe it's already implemented?

**Comment by eernstg** [comment]
No, that's a slightly different thing: The statement `a |= true` uses the [operator `|`](https://api.dartlang.org/stable/2.4.1/dart-core/bool/operator_bitwise_or.html) on `bool`, which is a regular operator. So expression evaluation for `e1 | e2` is a regular instance method invocation. In particular, `e2` _will_ be evaluated even in the case where `e1` evaluates to `true`.

This matters in practice, because you can use `done ||= doWork()` to skip the execution of `doWork()` in the case where `done` is already true.

**Comment by thosakwe** [comment]
Ah, I see.

**Comment by marcglasberg** [comment]
Hey there, was this abandoned? 

If I have this: 

```dart
onPressed: doSomething,
```

I'd like to be able to do:

```dart
onPressed: doSomething && doSomethingElse || doSomethingElseEntirely,
```

By defining something like:

```dart
extension FunctionExtension on bool Function(String) {
  bool Function(String) operator &&=(bool Function(String) other) =>
      (String value) => this(value) || other(value);

  bool Function(String) ||=(bool Function(String) other) =>
      (String value) => this(value) && other(value);
}
```

**Comment by eernstg** [comment]
It is not abandoned, cf. https://github.com/dart-lang/language/issues/1077. However, you cannot declare an instance operator for any compound assignment (`+=`, `*=`, etc.), and also not for `&&` or `||` (because evaluation of `e1 && e2` is different from a method invocation). But you can do this:

```dart
typedef F = bool Function(String);

extension LiftBool on F {
  F operator &(F other) => (s) => this(s) && other(s);
  F operator |(F other) => (s) => this(s) || other(s);
}

bool succeed(String s) { print('Succeed: $s'); return true; }
bool fail(String s) { print('Fail: $s'); return false; }
bool notReached(String s) => throw 'Not reached';

void main() {
  (succeed & succeed | notReached)('Hello');
  (fail & notReached | succeed)('world!');
}
```

**Comment by munificent** [comment]
I'm going to mark this as a "feature" and not a "request" because this issue really isn't describing a user problem, it's just a feature. I don't think the separate "request"/"feature" pairing really helps here. That's useful when the user issue is open-ended enough that multiple possible language features could solve the problem. But I think there's really only one way to solve "I want to write `&&=` and Dart won't let me." :)

**Comment by munificent** [comment]
Note that Lasse has a proposed specification for this feature here: https://dart-review.googlesource.com/c/sdk/+/43500.
